_drafts/Firmware-encryption-with-python.md

```diff
@@ -0,0 +1,643 @@
+
+Before moving on to implementation, let’s review some important points:
+
+1. CTR mode requires a key `(AES_KEYRx (KEY))` and a Nonce initialization vector `(AES_IVRx)`.
+2. We can encrypt plaintext data (AES_DINR (plaintext P1)) in blocks of various sizes, for example, 4 bytes at a time.
+3. Each time we encrypt a plaintext block, the counter block increases by one unit.
+4. The counter block `(AES_IVRx)` is the concatenation of a fixed Nonce block, set during initialization, and a 32-bit counter variable.
+
+By understanding these key points, we can proceed with the implementation of CTR mode encryption.
+
+## Firmware Encryption Implementation
+
+To begin with the implementation, let’s use the PyCryptodome cryptographic library to create a Python script that takes our firmware file and generates a new encrypted output file. First, let’s install the PyCryptodome library in Python. You can install it by running the following command:
+```
+$ pip install pycryptodome
+```
+
+### Define encryption keys
+
+To define the encryption keys, we will generate a key to use in the script using the PyCryptodome library. This same key will also be used for decryption in the MCU. For our configuration, we will use AES-128, which requires a 16-byte key. Additionally, we will use a 2-byte Nonce value as the initialization vector.
+
+```python
+from Crypto import Random
+myKey = Random.get_random_bytes(16)
+myIV  = Random.get_random_bytes(2)
+print(myKey.hex())
+print(myIV.hex())
+```
+Here are the generated keys:
+
+```
+$ 583bf4901f623510a1408e8be7bd6f5a
+$ f824
+```
+
+### Encrypting with PyCryptodome
+
+Let’s use the generated keys to start encrypting data. The first step is to configure our counter block, which is a necessary parameter to create an object that allows us to encrypt using the `.encrypt()` method.
+
+#### Counter Block
+
+According to the PyCryptodome documentation, we can create a counter block by concatenating values before (prefix) and after (suffix) the counter variable:
+
+![](/img/fwup-encryption/Coun_block_1.png)
+
+Let’s use our key and Nonce value (AES_IVRx) to create the counter block using the `Counter.new()` method. The method requires several parameters:
+
+1. The desired counter length in bits.
+2. The initial value of the counter.
+3. The endianness of the counter data alignment.
+
+> counter length: 112 bit / 8 = 14 bytes
+> counter block length: 14 bytes + Nonce 2 bytes = 16 bytes = 128 bit
+
+Since we want a counter length of 112 bits and our Nonce is 2 bytes, the counter block length should be 16 bytes (128 bits) to match the AES-128 bit encryption. We want the counter variable to start at 0, so we’ll set initial_value=0. We’ll also specify little_endian=False for big-endian data alignment to align with the counter in the AES processor on the MCU:
```

Comment:
```suggestion
Since we want a counter length of 112 bits and our Nonce is 2 bytes, the counter block length should be 16 bytes (128 bits) to match the AES-128 bit encryption. We want the counter variable to start at 0, so we’ll set initial_value=0. We’ll also specify `little_endian=False` for big-endian data alignment to align with the counter in the AES processor on the MCU:
```